modules/internal/ChapelDomain.chpl

```diff
@@ -462,14 +460,35 @@ module ChapelDomain {
 
   // any combinations not handled by the above
 
-  inline operator ==(d1: domain, d2: domain) param {
-    return false;
+  private proc cmpError(d1, d2) {
+    if d1.isRectangular() then
+      compilerError("comparing a rectangular domain against",
+        " an associative or sparse domain is not currently supported");
+    if d1.isAssociative() then
+      compilerError("comparing an associative domain against",
+        " a rectangular or sparse domain is not currently supported");
+    if d1.isSparse() then
+      compilerError("comparing a sparse domain against",
+        " a rectangular or associative domain is not currently supported");
+    // we should not get here, however if we do...
```

Comment:
Why not make this an assertion instead so it becomes evident that we have a bug in our module code?

Reply:
The `==` and `!=` operators are available to the public, so the error case is not necessarily our bug. Really, it may not be a bug at all. Hence a user-friendly error message.